misk/src/main/kotlin/misk/digester/MergingDigest.kt

```diff
@@ -0,0 +1,351 @@
+    var nextIndex = indexEstimate((beforeWeight + next.weight) / totalWeight)
+
+    if (nextIndex-beforeIndex > 1 || mainCentroids.size == 0) {
+      // the new index is far away from the last index of the current centroid
+      // therefore we cannot merge into the current centroid or it would
+      // become too wide, so we will append a new centroid
+      mainCentroids.add(next)
+      // return the last index that was merged into the previous centroid
+      return indexEstimate(beforeWeight / totalWeight)
+    } else {
+      // the new index fits into the range of the current centroid, so we
+      // combine it into the current centroid's values
+      // this computation is known as welford's method, the order matters
+      // weight must be updated before mean
+
+      //TODO: is this thread safe? doublecheck that this successfully updates all properties
+      var mainCentroidsTemp = mainCentroids[mainCentroids.size-1]
+      var builder = mainCentroidsTemp.newBuilder()
+      builder.weight += next.weight
+      builder.mean +=
+          (next.mean - mainCentroids[mainCentroids.size-1].mean) * next.weight / mainCentroids[mainCentroids.size-1].weight
+
+      // we did not create a new centroid, so the trailing index of the previous
+      // centroid remains
+      return beforeIndex
+    }
+  }
+
+  /**
+   * given a quantile, estimate the index of the centroid that contains it using
+   * the given compression
+   */
+  fun indexEstimate(quantile: Double) : Double {
+    return compression * ((asin(2*quantile-1) / PI) + 0.5).toFloat()
+  }
+
+  /**
+   * Returns a value such that the fraction of values in td below that value is
+   * approximately equal to quantile. Returns NaN if the digest is empty.
+   */
+  fun quantile(quantile: Double): Double {
+    if (quantile < 0 || quantile > 1) {
+      throw IndexOutOfBoundsException("quantile out of bounds")
+    }
+    mergeAllTemps()
+
+    // add up the weights of centroids in ascending order until we reach a
+    // centroid that pushes us over the quantile
+    var q = quantile * mainWeight
+    var weightSoFar = 0.0
+    var lowerBound = min
+
+    val iterator = mainCentroids.iterator()
+
+    for((i, c) in iterator.withIndex()) {
+      var upperBound = centroidUpperBound(i)
+      if (q <= weightSoFar+c.weight) {
+        // the target quantile is somewhere inside this centroid
+        // we compute how much of this centroid's weight falls into the quantile
+        var proportion = (q - weightSoFar) / c.weight
+        // and interpolate what value that corresponds to inside a uniform
+        // distribution
+        return lowerBound + (proportion * (upperBound - lowerBound))
+      }
+
+      // the quantile is above this centroid, so sum the weight and carry on
+      weightSoFar += c.weight
+      lowerBound = upperBound
+    }
+
+    // should never be reached unless empty, since the final comparison is
```

Comment:
if you should "never" reach this, throw an exception

Reply:
I dont think that an exception should be the expected behaviour in this case. Calling quantile when there is no populated data should still run and be valid, but should return NaN.